mqtt.go

```diff
@@ -20,147 +23,204 @@ const (
-		command        chan<- *DeviceCommand
-		event          <-chan *DeviceEvent
-		bulkData       <-chan *DeviceBulkData
-		syncedBulkData <-chan *DeviceSyncedBulkData
-		kvSync         chan<- *keyValueSync
-		kvPush         <-chan *keyValueSync
-		err            chan error
-		doPing         chan time.Duration
-		outPacket      chan packet.Packet
-		puback         chan *packet.PubackPacket
-		pingresp       chan *packet.PingrespPacket
-		stopPublisher  chan bool
-		wgWrite        sync.WaitGroup
-		wgRead         sync.WaitGroup
+	// For implementation dependent MQTT.
+	MqttDelegate interface {
+		// Returns authentication information
+		AuthInfo() (username string, password string)
+		// Builds the publish packet based on data in the interface.
+		BuildPublishPacket(data interface{}) (packet.Packet, error)
+		Subscriptions() []MqttSubscription
 	}
-)
 
-func (mc *mqttConn) close() {
-	close(mc.stopPublisher) // tell publisher to quit
-	close(mc.doPing)        // tell pinger to quit
+	MqttClient struct {
+		// Provides the public API to MQTT. We handle
+		// connect, disconnect, ping (internal)
+		// publish, and subscribe. All functions are synchronous.
 
-	mc.wgWrite.Wait() // wait for event processor and pinger to finish
+		isConnected  bool
+		delegate     MqttDelegate  // mqtt conn's need user/pass
+		deadline     time.Duration // our timeout duration
+		pingInterval time.Duration // our timeout duration
+		conn         *mqttConn
+		subs         map[string]MqttSubscription // map of subs to handlers
+		pingCtrl     chan bool                   // Channel to stop pings
+		wgSend       sync.WaitGroup              // wait on sending
+		wgRecv       sync.WaitGroup              // wait on receiving
+	}
 
-	// Attempt graceful disconnect.
-	mc.outPacket <- packet.NewDisconnectPacket()
-	close(mc.outPacket)
+	// Delegate for the connection to call back to the client
+	mqttPubReceiver interface {
+		handlePub(*packet.PublishPacket)
+	}
 
-	mc.wgRead.Wait() // wait for packet reader to finish
-}
+	// Type alias for something that we use everywhere
+	responseChan chan packet.Packet
 
-func (mc *mqttConn) getErrorChan() chan error {
-	return mc.err
-}
+	// Internal structure used by the client to communicate with the mqttd
+	// server
+	mqttConn struct {
+		PubHandler mqttPubReceiver // delegate to handle publish
 
-func (mc *mqttConn) ping(timeout time.Duration) {
-	mc.doPing <- timeout
-}
+		conn   net.Conn
+		stream *packet.Stream
+		// Sequential packet ID. Used to match acks our actions (pub, sub)
+		// excluding connects and pings. We also don't have packet ID's for
+		// receiving pubs from our subscriptions because we didn't initiate them.
+		lastPacketID uint16
 
-func (mc *mqttConn) sendPacket(p packet.Packet) error {
-	if err := mc.stream.Write(p); err != nil {
-		logError("[MQTT] failed to send %s packet: %s", p.Type(), err.Error())
-		return err
+		// Channels to respond to clients
+		packIDToChan map[uint16]responseChan
```

Comment:
I like this way. It can concurrency `PUBLISH`. I'm using similar way in FactoryGateway and HMI (Factory  controller) communication on MQTT.

I have a worrier. If a device works under the unstable network, maybe many fail to receive responses. Maybe it can delete when timeout (or `ctx.Done()`). But in this case, reasons of fail is many causes. 

1. It can not receive just PUBACK
2. disconnection in physical layer
3. Just delay. In this case, device receive PUBACK after timeout.  

Maybe difficult to distinguish `1.` and `2.`. 

But maybe the device can distinguish `3.`. In this case, there is no packerID in `packIDToChan` map.

So, I want you to that this SDK write detail to log if happened error.  Maybe customer ask me when unstable network. I hope that I can understand what happens under unstable network.

Reply:
Yes, I think it will block, but it doesn't handle errors too well. I discussed it with Ethan, and I will move the error handling outside of the mqtt API, and handle the error at a higher level. Thanks for your input!

Reply:
Oh... interesting > move the error handling outside of the mqtt API